File: Problem_2.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Take a hashmap to store the string to the index of both the word, and then traverse through the list of index of both words and get the min distance between the both
// Your code here along with comments explaining your approach
class WordDistance {

    HashMap<String, ArrayList<Integer>> hm;
    public WordDistance(String[] wordsDict) {
        hm = new HashMap<>();
        for(int i = 0; i < wordsDict.length; i++){
            if(!hm.containsKey(wordsDict[i])){
                hm.put(wordsDict[i], new ArrayList<>());
            }
            hm.get(wordsDict[i]).add(i);
        }
    }

    public int shortest(String word1, String word2) {
        ArrayList<Integer> a1 = hm.get(word1);
        ArrayList<Integer> a2 = hm.get(word2);
        int i = 0;
        int j = 0;
        int min = Integer.MAX_VALUE;
        while(i < a1.size() && j < a2.size()){
            min = Math.min(min, Math.abs(a1.get(i)-a2.get(j)));
            if(a1.get(i) < a2.get(j)){
                i++;
            }else{
                j++;
            }
        }
        return min;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */

File: Problem_2.js
// ## Problem 2: Minimum word distance ii (https://leetcode.com/problems/shortest-word-distance-ii/)

// TC: O(l)
// SC: O(nl), n is the number of words, l is average number of time it occurs

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

/**
 * @param {string[]} wordsDict
 */
var WordDistance = function (wordsDict) {
    let map = new Map();
    for (let i = 0; i < wordsDict.length; i++) {
        let word = wordsDict[i];
        if (!map.has(word)) {
            map.set(word, []);
        }
        map.get(word).push(i);
    }
    this.getMap = function () {
        return map;
    }
};

/** 
 * @param {string} word1 
 * @param {string} word2
 * @return {number}
 */
WordDistance.prototype.shortest = function (word1, word2) {
    let map = this.getMap();
    let l1 = map.get(word1);
    let l2 = map.get(word2);
    let min = Infinity;
    let p1 = 0, p2 = 0;
    while (p1 < l1.length && p2 < l2.length) {
        let val1 = l1[p1];
        let val2 = l2[p2];
        min = Math.min(min, Math.abs(val1 - val2))
        if (val1 < val2) {
            p1++;
        } else {
            p2++;
        }
    }
    return min;
};

/** 
 * Your WordDistance object will be instantiated and called as such:
 * var obj = new WordDistance(wordsDict)
 * var param_1 = obj.shortest(word1,word2)
 */
